addons/binding/org.openhab.binding.globalcache/src/main/java/org/openhab/binding/globalcache/handler/GlobalCacheHandler.java

```diff
@@ -788,10 +856,201 @@ private void checkConnection() {
+                if (StringUtils.isNotEmpty(eomString)) {
+                    logger.debug("End of message is {} for thing {} {}", eomString, thingID(), serialDevice);
+                    byte[] eom;
+                    try {
+                        eom = URLDecoder.decode(eomString, CHARSET).getBytes(CHARSET);
+                    } catch (UnsupportedEncodingException e) {
+                        logger.info("Unable to decode end of message delimiter {} for thing {} {}", eomString,
+                                thingID(), serialDevice, e);
+                        return null;
+                    }
+
+                    // Start the serial reader using the above end-of-message delimiter
+                    SerialReader serialReader = new SerialReader(serialDevice, getSerialIn(serialDevice), eom);
+                    serialReader.start();
+                    return serialReader;
+                } else {
+                    logger.warn("End of message delimiter is not defined in configuration of thing {}", thingID());
+                }
+            }
+            return null;
+        }
+
+        private void stopSerialReaders() {
+            if (deviceSupportsSerial1() && serialReader1 != null) {
+                serialReader1.stop();
+            }
+            if (deviceSupportsSerial2() && serialReader2 != null) {
+                serialReader2.stop();
+            }
+        }
+    }
+
+    /*
+     * The {@link SerialReader} class reads data from the serial connection. When data is
+     * received, the receive channel is updated with the data. Data is read up to the
+     * end-of-message delimiter defined in the Thing configuration.
+     *
+     * @author Mark Hilbush - Initial contribution
+     */
+    private class SerialReader {
+        private Logger logger = LoggerFactory.getLogger(SerialReader.class);
+
+        private CommandType serialPort;
+        private BufferedInputStream serialIn;
+        private ScheduledFuture<?> serialReaderJob;
+        private boolean terminateSerialReader;
+
+        private byte[] eom;
+
+        private Runnable serialReaderRunnable = new Runnable() {
+            @Override
+            public void run() {
+                serialReader();
+            }
+        };
+
+        SerialReader(CommandType serialPort, BufferedInputStream serialIn, byte[] eom) {
+            if (serialIn == null) {
+                throw new IllegalArgumentException("Serial input stream is not set");
+            }
+            this.serialPort = serialPort;
+            this.serialIn = serialIn;
+            this.eom = eom;
+            serialReaderJob = null;
+            terminateSerialReader = false;
+        }
+
+        public void start() {
+            serialReaderJob = scheduledExecutorService.schedule(serialReaderRunnable, 0, TimeUnit.SECONDS);
+        }
+
+        public void stop() {
+            if (serialReaderJob != null) {
```

Comment:
The streams are opened and closed in one place (serialConnect and serialDisconnect).